components/tikv_util/src/security.rs

```diff
@@ -137,6 +138,9 @@ impl SecurityManager {
         } else {
             let fetcher = Box::new(Fetcher {
                 cfg: self.cfg.clone(),
+                last_fetch: Arc::new(Mutex::new(
```

Comment:
should the `Mutex` be around the whole `Fetcher`, not just `last_fetch`?

Reply:
This mutex is only for restricting the modification of `last_fetch`. 
If I can add mutex directly to this fecher, it's better. But...This will force grpc-rs users who don't need mutex to use mutex. Maybe they only return `Ok(None)` most of the time. Another important reason is that the callback in grpc-core is not protected by mutex, so it is useless to lock in the callback function itself.

Reply:
If two caller calling `fetch` concurrently, they may both see a timeout and both go to fetch the config. If we lock the whole struct then one of them will wait while the other is fetching config.

Reply:
This is where I find rust confusing.. By the logic, `cfg` is essentially also behind the mutex, since all caller of `fetch` need to wait on the mutex before getting access `cfg`. However the syntax make it appear that only `last_fetch` is behind the mutex. IMHO have the mutex guard the whole struct make the related more clear.

Reply:
1. Only one thread can get lock and see a timeout.
2. Locking the entire fetcher has no effect, because grpc-core could handle the result of the callback function concurrently.

Reply:
Locking the entire Fetcher prevents connections that do not require a certificate update.

In the present case, they can return `Ok(None)` directly after the try fails, which means that they can continue to establish a connection using the previous certificate configuration

Reply:
> This is where I find rust confusing.. By the logic, `cfg` is essentially also behind the mutex, since all caller of `fetch` need to wait on the mutex before getting access to `cfg`. However the syntax make it appear that only `last_fetch` is behind the mutex. IMHO have the mutex guard the whole struct make the relation more clear.

@wujy-cs I know what you mean, but `cfg` is a read-only variable, adding mutex may cause ambiguity

Reply:
> Locking the entire Fetcher prevents connections that do not require a certificate update.
> 
> In the present case, they can return `Ok(None)` directly after the try fails, which means that they can continue to establish a connection using the previous certificate configuration

I didn't notice it is a `try_lock` instead of `lock`. Can you add some inline comment about the "return Ok(None) immediately" behavior?

Reply:
OK, done